File: bindings/python/tests/test_stream.py
"""Streaming encode tests for the Python binding.

Wraps the post-1.18 C ABI's `ztok_stream_*` family. The streaming output
should match `Pipeline.encode` byte-for-byte when the pre-tokenizer is a
real one (cl100k splits on whitespace, giving safe-cut boundaries on
every span) and may differ ONLY at unsafe chunk seams with the identity
pre-tokenizer + BPE-merge models — same caveat documented in
src/stream.zig.
"""

from __future__ import annotations

from pathlib import Path

import pytest

import ztok


def _collect(stream) -> list[int]:
    """Materialize a generator of id lists into a single flat list."""

    out: list[int] = []
    for batch in stream:
        assert isinstance(batch, list)
        out.extend(batch)
    return out


def test_stream_single_chunk_matches_encode(bpe_pipeline: ztok.Pipeline) -> None:
    text = "hello world the quick brown fox"
    want = bpe_pipeline.encode(text)
    got = _collect(bpe_pipeline.encode_stream(text))
    assert got == want


def test_stream_many_small_chunks_matches_encode(bpe_pipeline: ztok.Pipeline) -> None:
    # Force lots of feeds by setting a tiny chunk size — the encoder
    # should still emit ids identical to a single-shot encode because
    # the cl100k pre-tokenizer finds safe whitespace cuts on every span.
    text = "hello world the quick brown fox hello world the quick brown fox"
    want = bpe_pipeline.encode(text)
    got = _collect(bpe_pipeline.encode_stream(text, chunk_size=4))
    assert got == want


def test_stream_mid_utf8_codepoint_is_deferred() -> None:
    # "héllo" — the é is 2 bytes (0xC3 0xA9). Feeding the bytes in two
    # halves where the cut falls between the two bytes of é must not
    # produce a garbage id; the stream encoder defers the trailing 0xC3
    # to the next feed.
    with ztok.Pipeline.byte_id() as pipe:
        text = "h\xe9llo"
        # bytes manually to control the cut precisely.
        raw = text.encode("utf-8")
        want = pipe.encode(text)

        # Drive the underlying C API directly to control the cut point.
        # encode_stream(chunk_size=2) chops at byte 2 (mid-é).
        got = _collect(pipe.encode_stream(text, chunk_size=2))
        assert got == want
        assert len(raw) == 6  # "h" + 0xC3 0xA9 + "llo"


def test_stream_empty_input_yields_nothing() -> None:
    with ztok.Pipeline.byte_id() as pipe:
        out = list(pipe.encode_stream(""))
        # Empty input produces zero non-empty yields.
        assert out == []


def test_stream_multiline_matches_encode(bpe_pipeline: ztok.Pipeline) -> None:
    text = "hello world\nthe quick brown fox\nfoo bar baz\n"
    want = bpe_pipeline.encode(text)
    got = _collect(bpe_pipeline.encode_stream(text, chunk_size=8))
    assert got == want

File: bindings/ruby/test/test_stream.rb
# frozen_string_literal: true

require_relative "helper"

class TestStream < Minitest::Test
  include ZtokTestFixtures

  def setup
    @pipe = Ztok::Pipeline.from_tiktoken(ZtokTestFixtures.tiktoken_fixture, cl100k: true)
  end

  def teardown
    @pipe&.close
  end

  def collect(enum)
    out = []
    enum.each do |batch|
      assert_kind_of Array, batch
      out.concat(batch)
    end
    out
  end

  def test_stream_single_chunk_matches_encode
    text = "hello world the quick brown fox"
    want = @pipe.encode(text)
    got = collect(@pipe.encode_stream(text))
    assert_equal want, got
  end

  def test_stream_many_small_chunks_matches_encode
    text = "hello world the quick brown fox hello world the quick brown fox"
    want = @pipe.encode(text)
    got = collect(@pipe.encode_stream(text, chunk_size: 4))
    assert_equal want, got
  end

  def test_stream_mid_utf8_codepoint_is_deferred
    # "héllo" — é is 0xC3 0xA9 (2 bytes). chunk_size=2 chops mid-é.
    Ztok::Pipeline.byte_id do |bpipe|
      text = "héllo"
      want = bpipe.encode(text)
      got = collect(bpipe.encode_stream(text, chunk_size: 2))
      assert_equal want, got
      assert_equal 6, text.bytesize # "h" + 0xC3 0xA9 + "llo"
    end
  end

  def test_stream_empty_input_yields_nothing
    Ztok::Pipeline.byte_id do |bpipe|
      out = bpipe.encode_stream("").to_a
      assert_equal [], out
    end
  end

  def test_stream_multiline_matches_encode
    text = "hello world\nthe quick brown fox\nfoo bar baz\n"
    want = @pipe.encode(text)
    got = collect(@pipe.encode_stream(text, chunk_size: 8))
    assert_equal want, got
  end

  def test_stream_encoder_low_level
    text = "hello world the quick brown fox"
    want = @pipe.encode(text)

    encoder = Ztok::StreamEncoder.new(@pipe)
    got = []
    begin
      # Feed two halves with arbitrary cut.
      mid = text.bytesize / 2
      got.concat(encoder.feed(text.byteslice(0, mid)))
      got.concat(encoder.feed(text.byteslice(mid, text.bytesize - mid)))
      got.concat(encoder.finish)
    ensure
      encoder.close
    end
    assert_equal want, got
  end
end
